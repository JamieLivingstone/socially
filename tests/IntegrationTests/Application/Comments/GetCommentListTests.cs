using System.Linq;
using System.Threading.Tasks;
using Application.Comments.Queries.GetCommentList;
using Application.Common.Exceptions;
using FluentAssertions;
using IntegrationTests.Application.TestUtils;
using NUnit.Framework;
using Snapshooter.NUnit;
using ValidationException = FluentValidation.ValidationException;

namespace IntegrationTests.Application.Comments;

[TestFixture]
public class GetCommentListTests : TestBase
{
  [Test]
  public void GivenAnInvalidRequest_ThrowsValidationException()
  {
    var query = new GetCommentListQuery
    {
      Slug = Seed.Posts().First().Slug,
      PageNumber = 0, // Must be 1 or greater
      PageSize = 10
    };

    FluentActions.Invoking(() => SendAsync(query)).Should().ThrowAsync<ValidationException>();
  }

  [Test]
  public void GivenPostDoesNotExist_ThrowsNotFoundException()
  {
    var query = new GetCommentListQuery
    {
      Slug = "does-not-exist",
      PageNumber = 1,
      PageSize = 10
    };

    async Task Handler() => await SendAsync(query);

    Assert.ThrowsAsync(typeof(NotFoundException), Handler);
  }

  [Test]
  public async Task GivenAValidRequest_ReturnsPaginatedComments()
  {
    var pageOne = await SendAsync(new GetCommentListQuery
    {
      Slug = Seed.Posts().First().Slug,
      PageNumber = 1,
      PageSize = 3
    });

    var pageTwo = await SendAsync(new GetCommentListQuery
    {
      Slug = Seed.Posts().First().Slug,
      PageNumber = 2,
      PageSize = 3
    });

    Snapshot.Match(new
    {
      pageOne,
      pageTwo
    });
  }
}
