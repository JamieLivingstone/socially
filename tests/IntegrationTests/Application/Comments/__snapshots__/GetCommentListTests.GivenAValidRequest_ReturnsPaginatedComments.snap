{
  "pageOne": {
    "Items": [
      {
        "Id": 1,
        "Message": "Mock comment one",
        "CreatedAt": "0001-01-01T00:00:00",
        "Author": {
          "Name": "Liam Hale",
          "Username": "liam"
        }
      },
      {
        "Id": 2,
        "Message": "Mock comment two",
        "CreatedAt": "0001-01-01T00:00:00",
        "Author": {
          "Name": "Liam Hale",
          "Username": "liam"
        }
      },
      {
        "Id": 3,
        "Message": "Mock comment three",
        "CreatedAt": "0001-01-01T00:00:00",
        "Author": {
          "Name": "Peter Gray",
          "Username": "peter"
        }
      }
    ],
    "PageNumber": 1,
    "TotalPages": 2,
    "TotalCount": 4,
    "HasPreviousPage": false,
    "HasNextPage": true
  },
  "pageTwo": {
    "Items": [
      {
        "Id": 4,
        "Message": "Mock comment four",
        "CreatedAt": "0001-01-01T00:00:00",
        "Author": {
          "Name": "Peter Gray",
          "Username": "peter"
        }
      }
    ],
    "PageNumber": 2,
    "TotalPages": 2,
    "TotalCount": 4,
    "HasPreviousPage": true,
    "HasNextPage": false
  }
}
